metaswitch/common/test/test_alarms.py

```diff
@@ -0,0 +1,302 @@
+    def test_multi_get_alarm(self, mock_atexit, mock_condition):
+        """We can get a multi-severity alarm from the manager."""
+        alarm_manager = _AlarmManager()
+
+        # We don't want to start it re-sending alarms.
+        with mock.patch.object(alarm_manager, 'start') as mock_start:
+            alarm = alarm_manager.get_alarm('TestIssuer', (1000, 1, 2, 6))
+
+        # Check that the alarm works as expected.
+        alarm.set(6)
+        alarm.clear()
+
+        # Check that the alarm_manager tried to start and registered for
+        # cleanup.
+        mock_start.assert_called_once_with()
+        mock_atexit.register.assert_called_once_with(alarm_manager.terminate)
+
+    @mock.patch('threading.Condition', autospec=True)
+    @mock.patch('metaswitch.common.alarms.atexit', autospec=True)
+    def test_bad_alarm(self, mock_atexit, mock_condition):
+        """The alarm manager doesn't create bad alarms."""
+        alarm_manager = _AlarmManager()
+
+        # We don't want to start it re-sending alarms.
+        with mock.patch.object(alarm_manager, 'start') as mock_start:
+            # No data.
+            self.assertRaises(Exception,
+                              alarm_manager.get_alarm,
+                              'TestIssuer',
+                              ())
+            # No severities.
+            self.assertRaises(Exception,
+                              alarm_manager.get_alarm,
+                              'TestIssuer',
+                              (1000))
+            # Only cleared severity.
+            self.assertRaises(Exception,
+                              alarm_manager.get_alarm,
+                              'TestIssuer',
+                              (1000, CLEARED))
+
+
+class TestAlarmManagerReSync(unittest.TestCase):
+
+    @mock.patch('metaswitch.common.alarms._sendrequest')
+    @mock.patch('metaswitch.common.alarms.atexit', autospec=True)
+    def test_terminate(self, mock_atexit, mock_sendrequest):
+        """Check that alarm manager termination takes under 5s."""
+        alarm_manager = _AlarmManager()
+        alarm_manager.start()
+
+        terminate_thread = threading.Thread(target=alarm_manager.terminate)
+
+        # Don't let the test hang due to this thread remaining.
+        terminate_thread.daemon = True
+        terminate_thread.start()
+
+        # Give the manager time to get started.
+        time.sleep(0.1)
+
+        terminate_thread.join(5)
+        self.assertFalse(terminate_thread.is_alive())
+
+    @mock.patch('metaswitch.common.alarms.monotonic')
+    @mock.patch('metaswitch.common.alarms._sendrequest')
+    @mock.patch('metaswitch.common.alarms.atexit', autospec=True)
+    def test_resync(self, mock_atexit, mock_sendrequest, mock_monotonic):
+        """Check that the alarm manager sends alarms after 30s."""
+        try:
+            # Start at 0 during initialization.
+            # Check that we account for the 5s delay in getting started.
```

Comment:
I've written the code to cope with a slow start, so this tests that it does. 5s isn't very realistic, but it could happen e.g. during a VMware vMotion to a new host.